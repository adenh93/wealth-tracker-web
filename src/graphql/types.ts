export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CryptoCurrency = {
  __typename?: 'CryptoCurrency';
  id: Scalars['Int'];
  name: Scalars['String'];
  rank: Scalars['Int'];
  symbol: Scalars['String'];
};

export type CryptoCurrencyHolding = {
  __typename?: 'CryptoCurrencyHolding';
  cryptoCurrency?: Maybe<CryptoCurrency>;
  cryptoCurrencyId: Scalars['Int'];
  holdings: Scalars['Float'];
  id: Scalars['Int'];
  netHoldingsValue: Scalars['Float'];
  percentChange24h: Scalars['Float'];
  price: Scalars['Float'];
};

export type CryptoCurrencyHoldingsAddInput = {
  cryptoCurrencyId: Scalars['Int'];
  holdings: Scalars['Float'];
};

export type CryptoCurrencyHoldingsUpdateInput = {
  holdings: Scalars['Float'];
  id: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  cryptoCurrencyHoldingsAdd?: Maybe<CryptoCurrencyHolding>;
  cryptoCurrencyHoldingsDelete?: Maybe<CryptoCurrencyHolding>;
  cryptoCurrencyHoldingsUpdate?: Maybe<CryptoCurrencyHolding>;
};


export type MutationCryptoCurrencyHoldingsAddArgs = {
  input: CryptoCurrencyHoldingsAddInput;
};


export type MutationCryptoCurrencyHoldingsDeleteArgs = {
  id: Scalars['Int'];
};


export type MutationCryptoCurrencyHoldingsUpdateArgs = {
  input: CryptoCurrencyHoldingsUpdateInput;
};

export type Query = {
  __typename?: 'Query';
  cryptoCurrencies?: Maybe<Array<CryptoCurrency>>;
  cryptoCurrencyHoldings?: Maybe<Array<CryptoCurrencyHolding>>;
};


export type QueryCryptoCurrenciesArgs = {
  query?: Maybe<Scalars['String']>;
};
