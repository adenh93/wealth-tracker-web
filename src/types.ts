export interface Asset {
  id: string
  logoSrc: string
  name: string
  ticker: string
}

export interface OwnedAsset extends Asset {
  price: number
  gain24Hr: number
  holdings: number
}
