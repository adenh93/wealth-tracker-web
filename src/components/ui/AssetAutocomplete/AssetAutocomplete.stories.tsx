import { Story } from '@storybook/react'
import AssetAutocomplete, { AssetAutocompleteProps } from './AssetAutocomplete'
import { Asset } from '../../../types'

const config = {
  title: 'Components/UI/AssetAutocomplete',
  component: AssetAutocomplete,
}

const Template: Story<AssetAutocompleteProps> = (args) => (
  <AssetAutocomplete {...args} />
)

export const Primary = Template.bind({})

const options: Asset[] = [
  {
    id: 'test-id-1',
    logoSrc: 'http://simpleicon.com/wp-content/uploads/apple.png',
    name: 'Apple Inc.',
    ticker: 'AAPL',
  },
  {
    id: 'test-id-2',
    logoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/4/4e/BitConnect.png',
    name: 'Bitconnect',
    ticker: 'BCC',
  },
  {
    id: 'test-id-3',
    logoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/b/bb/Tesla_T_symbol.svg/1200px-Tesla_T_symbol.svg.png',
    name: 'Tesla',
    ticker: 'TSLA',
  },
  {
    id: 'test-id-4',
    logoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/BTC_Logo.svg/1200px-BTC_Logo.svg.png',
    name: 'Bitcoin',
    ticker: 'BTC',
  },
  {
    id: 'test-id-4',
    logoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/4/44/Microsoft_logo.svg/512px-Microsoft_logo.svg.png',
    name: 'Microsoft',
    ticker: 'MSFT',
  },
]

Primary.args = {
  options,
  onChange: () => {},
}

export default config
