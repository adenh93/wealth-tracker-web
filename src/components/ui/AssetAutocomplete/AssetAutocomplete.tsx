import { FC } from 'react'
import { Autocomplete, Box, TextField } from '@mui/material'
import { Asset } from '../../../types'

export interface AssetAutocompleteProps {
  options: Asset[]
  error?: string
  onChange: (e: any, asset: Asset | null) => void
}

const AssetAutocomplete: FC<AssetAutocompleteProps> = ({
  options,
  error,
  onChange,
}) => {
  const getOptionLabel = (option: Asset): string =>
    `${option.name} (${option.ticker})`

  return (
    <Autocomplete
      autoHighlight
      size="small"
      options={options}
      onChange={onChange}
      getOptionLabel={getOptionLabel}
      renderOption={(props, option: Asset) => (
        <Box
          component="li"
          sx={{ '& > img': { mr: 2, flexShrink: 0 } }}
          {...props}
        >
          <img
            loading="lazy"
            width="20"
            src={option.logoSrc}
            alt={`${option.name} logo`}
          />
          {option.name} ({option.ticker})
        </Box>
      )}
      renderInput={(params) => (
        <TextField
          {...params}
          label="Choose an Asset"
          inputProps={{
            ...params.inputProps,
            autoComplete: 'new-password',
          }}
          error={!!error}
          helperText={error}
        />
      )}
    />
  )
}

export default AssetAutocomplete
