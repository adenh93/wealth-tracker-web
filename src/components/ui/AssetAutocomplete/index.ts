export { default } from './AssetAutocomplete'
export type { AssetAutocompleteProps } from './AssetAutocomplete'