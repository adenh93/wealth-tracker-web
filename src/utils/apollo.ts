import { ApolloError } from '@apollo/client'

export const extractInputErrorField = <FormType>(
  error: ApolloError
): keyof FormType | null => {
  if (error.graphQLErrors.length)
    return error.graphQLErrors[0].extensions?.field as keyof FormType
  return null
}
