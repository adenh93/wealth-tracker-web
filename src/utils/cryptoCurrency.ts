import { CMC_LOGO_URL } from '../constants'

export const getCMCLogoUrl = (id: number): string => `${CMC_LOGO_URL}/${id}.png`
