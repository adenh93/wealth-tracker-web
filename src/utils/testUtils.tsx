import { ReactElement } from 'react'
import { ThemeProvider } from '@mui/material'
import { render as tlRender, RenderResult } from '@testing-library/react'
import theme from '../theme'

export const render = (component: ReactElement): RenderResult =>
  tlRender(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
