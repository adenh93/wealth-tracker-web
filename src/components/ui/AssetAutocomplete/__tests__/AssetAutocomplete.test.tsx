import { Asset } from '../../../../types'
import { render } from '../../../../utils/testUtils'
import AssetAutocomplete from '../AssetAutocomplete'

describe('AssetAutocomplete component tests', () => {
  const mockOptions: Asset[] = [
    {
      id: 'test-id-1',
      logoSrc: 'http://simpleicon.com/wp-content/uploads/apple.png',
      name: 'Apple Inc.',
      ticker: 'AAPL',
    },
    {
      id: 'test-id-2',
      logoSrc:
        'https://upload.wikimedia.org/wikipedia/commons/4/4e/BitConnect.png',
      name: 'Bitconnect',
      ticker: 'BCC',
    },
  ]

  test('matches snapshot', () => {
    const { asFragment } = render(
      <AssetAutocomplete options={mockOptions} onChange={jest.fn()} />
    )

    expect(asFragment).toMatchSnapshot()
  })
})
