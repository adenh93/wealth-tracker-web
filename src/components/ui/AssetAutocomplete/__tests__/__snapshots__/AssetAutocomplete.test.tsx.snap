// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetAutocomplete component tests matches snapshot 1`] = `[Function]`;
